docs/src/appendix/mast.md

```diff
@@ -0,0 +1,528 @@
+
+The downside is that the tree structure is no longer enforced by the type system, instead we must maintain those invariants ourselves. This is a tradeoff, but one we've determined is worth making.
+
+Next, let's take a closer look at the definition of `MastNode` and the handle type used to reference `MastNode`s associated with a specific instance of `MastForest`:
+
+```rust
+/// We're using this simple type alias to reflect the fact that `MastNodeId`
+/// is ultimately an index into the `nodes` vector of a `MastForest`. However,
+/// in practice this is a more complex type that provides us with some
+/// additional guarantees/safety, and guards against accidental misuse.
+pub type MastNodeId = usize;
+
+pub enum MastNode {
+    /// See note following this snippet
+    Block(BasicBlockNode),
+    Join(JoinNode),
+    Split(SplitNode),
+    Loop(LoopNode),
+    Call(CallNode),
+    Dyn(DynNode),
+    /// A reference to a node whose definition is not
+    /// local to the containing `MastForest`.
+    External(Digest),
+}
+```
+
+> [!NOTE]
+> I'm using the "basic block" nomenclature here for clarity, and to better reflect its relationship to the semantics of basic blocks as they are known in standard compiler IRs and assemblers.
+>
+> In the current `miden-core` code, "block" is used for many things that are not actually blocks, or even block-like. This in turn seems to have forced the use of "span" to represent basic blocks. Since we've been starting to favor a more graph-theoretic set of terms for MAST in recent discussions, e.g. "tree"/"forest"/"node", I think it is clearer if we discard the previous usage of "block" in favor of "node", and "span" in favor of "basic block".
+
+The definitions of the concrete node types largely match what exists today, but we will look at them anyway just to illustrate any differences.
+
+First, a few preliminaries. We reference two traits which currently do not exist, but which represent common functionality across node types. We assume that they are derivable for purposes of this document:
+
+```rust
+pub trait MerkleTreeNode {
+    fn digest(&self) -> Digest;
+}
+
+pub trait Decorated {
+    fn decorators(&self) -> &DecoratorList;
+    fn decorate(&mut self, decorators: DecoratorList);
+}
+```
+
+Let's start with the _basic block_, which is what comprises the majority of leaf nodes in the MAST:
+
+```rust
+#[derive(MerkleTreeNode, Decoratable)]
+pub struct BasicBlockNode {
+    /// The primitive operations contained in this basic block.
+    ///
+    /// The operations are broken up into batches of 8 groups,
+    /// with each group containing up to 9 operations, or a
+    /// single immediates. Thus the maximum size of a basic block
+    /// is 72 operations. A larger block would need to be encoded
+    /// in MAST as a join of two basic blocks.
+    batches: Vec<OpBatch<8>>,
+}
+
+pub struct OpBatch<const BATCH_SIZE: usize = 8> {
+    ops: Vec<Operation>,
+    groups: [Felt; BATCH_SIZE],
```

Comment:
The primary reason for this is the way the VM decodes programs. Since `Felt` is the smallest unit the VM understands natively, we encode up to 9 operations into a single `Felt` and we call these an "operation group". Then, up to 8 groups can be absorbed by a single permutation of the hash function we use (RPO) and the VM needs to execute additional operations between decoding more than 8 groups (i.e., a single batch). At the very high level, when we execute a basic block, the VM actually executes the following operations (the basic block consists of 3 batches):
```
SPAN
<operations from batch 0>
RESPAN
<operations from batch 1>
RESPAN
<operations from batch 2>
END
```

Reply:
Got it. Thanks for the detailed explanation!

Reply:
I was also confused by this when reviewing the implementation of things in `miden-core`, but after re-reading things several times I was able to arrive at more or less what @bobbinth explained here. As an aside, your explanation here is actually better, and more succinct than anything I came across, we might actually want to add it to the code in `miden-core` as a brief summary for future readers!